pkg/scheduler/scheduler.go

```diff
@@ -319,6 +319,32 @@ func (s *Scheduler) nominate(ctx context.Context, workloads []workload.Info, sna
 	return entries
 }
 
+// resourcesToReserve calculates how much of available resources in cq assignment should be reserved.
+func resourcesToReserve(e *entry, cq *cache.ClusterQueue) cache.FlavorResourceQuantities {
+	if e.assignment.RepresentativeMode() != flavorassigner.Preempt {
+		return e.assignment.Usage
+	}
+	reservedUsage := make(cache.FlavorResourceQuantities)
+	for flavor, resourceUsage := range e.assignment.Usage {
+		cqQuota := make(map[corev1.ResourceName]*cache.ResourceQuota)
+		for _, rg := range cq.ResourceGroups {
+			for _, cqFlavor := range rg.Flavors {
+				if cqFlavor.Name == flavor {
+					cqQuota = cqFlavor.Resources
+				}
+			}
+		}
+		for resource, usage := range resourceUsage {
+			if _, ok := reservedUsage[flavor]; !ok {
+				reservedUsage[flavor] = make(map[corev1.ResourceName]int64)
+			}
+			// since assignment doesn't borrow, it's usage can't exceed nominal quota.
+			reservedUsage[flavor][resource] = min(usage, cqQuota[resource].Nominal-cq.Usage[flavor][resource])
```

Comment:
If the CQ is currently borrowing, Nominal-Usage is negative.

Add a unit test to cover this case.